numba/np/arrayobj.py

```diff
@@ -4545,21 +4531,28 @@ def impl(start, stop=None, step=None, dtype=None):
     return impl
 
 
-@glue_lowering(np.linspace, types.Number, types.Number)
-def numpy_linspace_2(context, builder, sig, args):
+@overload(np.linspace)
+def numpy_linspace_2(start, stop):
+    if isinstance(start, types.Number) and isinstance(stop, types.Number):
+        def impl(start, stop):
+            return np.linspace(start, stop, 50)
+        return impl
 
-    def linspace(start, stop):
-        return np.linspace(start, stop, 50)
 
-    res = context.compile_internal(builder, linspace, sig, args)
-    return impl_ret_new_ref(context, builder, sig.return_type, res)
+@overload(np.linspace)
+def numpy_linspace_3(start, stop, num):
+    if not all(isinstance(arg, types.Number) for arg in [start, stop]):
+        return
 
+    if not isinstance(num, types.Integer):
+        return
 
-@glue_lowering(np.linspace, types.Number, types.Number, types.Integer)
-def numpy_linspace_3(context, builder, sig, args):
-    dtype = as_dtype(sig.return_type.dtype)
+    if any(isinstance(arg, types.Complex) for arg in [start, stop]):
+        dtype = types.complex128
+    else:
+        dtype = types.float64
+    dtype = as_dtype(dtype)
```

Comment:
```suggestion
    if any(isinstance(arg, types.Complex) for arg in [start, stop]):
        dtype = np.complex128
    else:
        dtype = np.float64
```
Just set the dtype directly?